pkg/util/clusterdata/clusterdata.go

```diff
@@ -0,0 +1,129 @@
+package clusterdata
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+// TODO: After OpenShift 4.4, replace github.com/openshift/cluster-api with github.com/openshift/machine-api-operator
+import (
+	"context"
+	"sync"
+
+	azureproviderapis "github.com/openshift/cluster-api-provider-azure/pkg/apis"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/rest"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/util/restconfig"
+)
+
+var scheme = runtime.NewScheme()
+var codecs = serializer.NewCodecFactory(scheme)
+
+func init() {
+	utilruntime.Must(azureproviderapis.AddToScheme(scheme))
+}
+
+// OpenShiftClusterEnricher must update the cluster object with
+// data received from API server
+type OpenShiftClusterEnricher interface {
+	Enrich(ctx context.Context, ocs ...*api.OpenShiftCluster) error
+}
+
+type enricherFeatureConstructor func(*rest.Config, *api.OpenShiftCluster) (enricherFeature, error)
+type enricherFeature interface {
+	Name() string
+	FetchData(ctx context.Context) (func(), error)
+}
+
+// NewBestEffortEnricher returns an enricher that attempts to populate
+// fields, but ignores errors in case of failures
+func NewBestEffortEnricher(log *logrus.Entry, env env.Interface) OpenShiftClusterEnricher {
+	return &bestEffortEnricher{
+		log: log,
+		env: env,
+
+		restConfig: restconfig.RestConfig,
+		featureConstructors: []enricherFeatureConstructor{
+			newClusterVersionEnricherFeature,
+			newServicePrincipalEnricherFeature,
+			newWorkerProfilesEnricherFeature,
+		},
+	}
+}
+
+type bestEffortEnricher struct {
+	log *logrus.Entry
+	env env.Interface
+
+	restConfig          func(env env.Interface, oc *api.OpenShiftCluster) (*rest.Config, error)
+	featureConstructors []enricherFeatureConstructor
+}
+
+func (e *bestEffortEnricher) Enrich(ctx context.Context, ocs ...*api.OpenShiftCluster) error {
+	// TODO: Make use of the deadline from ctx
+	// NOTE: The above might need to wait for this KEP to land in openshift/client-go:
+	//       https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20200123-client-go-ctx.md
```

Comment:
Alternative options:

* Make a copy of the clients with the support of context. `k8s.io/client-go/rest.Request` seems to support contexts already, so we will need to define own clients implementing `configclient.Interface`, `kubernetes.Interface` and `clusterapi.Interface` with an extra method to set context.
* Less extra code: set `Timeout` on `k8s.io/client-go/rest.Config`. In this case each request will have own timeout. This might be acceptable because the time spread between requests should not be significant in the current implementation: we are currently starting requests concurently and not limiting amount of concurent requests. But it is quite fragile: if we change implementation details - the timeouts will start work not as expected.